File: app/src/main/java/com/slidely/photoframe/features/media/MediaViewModelFactory.java
package com.slidely.photoframe.features.media;

import androidx.lifecycle.ViewModel;
import androidx.lifecycle.ViewModelProvider;
import androidx.annotation.NonNull;

import com.slidely.photoframe.repository.MediaRepository;
import com.slidely.photoframe.repository.PreferenceRepository;

public class MediaViewModelFactory extends ViewModelProvider.NewInstanceFactory {
    private final MediaRepository mediaRepository;
    private final MediaView mediaView;
    private final PreferenceRepository preferenceRepository;

    MediaViewModelFactory(MediaView mediaView, MediaRepository mediaRepository,
                          PreferenceRepository preferenceRepository) {
        this.mediaView = mediaView;
        this.mediaRepository = mediaRepository;
        this.preferenceRepository = preferenceRepository;
    }

    @NonNull
    @Override
    public <T extends ViewModel> T create(@NonNull Class<T> modelClass) {
        return (T) new MediaViewModel(mediaView, mediaRepository, preferenceRepository);
    }
}

File: app/src/main/java/com/k4dima/photoframe/features/main/presentation/MediaViewModelFactory.kt
package com.k4dima.photoframe.features.main.presentation

import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import com.k4dima.photoframe.features.main.ui.di.MainScope
import javax.inject.Inject
import javax.inject.Provider

@MainScope
class MediaViewModelFactory
@Inject
internal constructor(private val creators: Map<Class<out ViewModel>, @JvmSuppressWildcards Provider<ViewModel>>) :
        ViewModelProvider.Factory {
    override fun <T : ViewModel> create(modelClass: Class<T>) =
            creators[modelClass]!!.get() as T
}